mlflow/gateway/__init__.py

```diff
@@ -0,0 +1,136 @@
+    """
+    server_state = read_server_state(server_state_file)
+
+    if server_state:
+        raise MlflowException(
+            f"There is a currently running server instance at pid '{server_state['pid']}'. Please "
+            "terminate the running server instance prior to starting a new instance within this "
+            "context.",
+            error_code=BAD_REQUEST,
+        )
+
+    os.environ[CONF_PATH_ENV_VAR] = config_path
+
+    cmd_path = pathlib.Path(__file__).parent.joinpath("gateway_app.py")
+    server_cmd = [sys.executable, str(cmd_path), "--host", host, "--port", str(port)]
+
+    server_process = subprocess.Popen(server_cmd)
+
+    url = f"http://{host}:{port}/health"
+    wait_until_server_starts(url=url, server_process=server_process, logger=_logger)
+
+    # Write the server state to disk
+    write_server_state(
+        host=host, port=port, pid=server_process.pid, server_state_file=server_state_file
+    )
+
+    return server_process
+
+
+def _stop_server(server_pid: int):
+    """
+    Stops the currently running server.
+
+    This function kills the process of the running server and resets the global server_process
+    variable to None.
+
+    Args:
+        server_state_file (str): The location on the local file system to store the server state
+
+    Returns:
+        None
+    """
+
+    if is_pid_alive(server_pid):
+        kill_parent_and_child_processes(server_pid)
+        _logger.info("The Gateway server has been terminated.")
+    else:
+        raise MlflowException(
+            "There is no currently running gateway server", error_code=BAD_REQUEST
+        )
+
+
+def _update_server(config_path: str, server_state_file: str = GATEWAY_SERVER_STATE_FILE):
+    """
+    Updates the server with the new configuration.
+
+    This function stops the currently running server and starts it again with the new configuration.
+    It raises an exception if there is no currently running server.
+
+    Args:
+        config_path (str): Path to the new configuration file.
+        server_state_file (str): The location on the local file system to store the server state
+
+    Returns:
+        None
+
+    Raises:
+        mlflow.exceptions.MlflowException: If there is no currently running server.
+    """
+    server_state = read_server_state(server_state_file)
+
+    if not server_state:
+        raise MlflowException(
+            f"There is no server_state_file present at {server_state_file}. Verify that the "
+            f"server has been started",
+            error_code=BAD_REQUEST,
+        )
+
+    server_pid = server_state["pid"]
+
+    if not is_pid_alive(server_pid):
+        raise MlflowException(
+            f"Unable to update server configuration. There is no server currently running at "
+            f"process id {server_pid}",
+            error_code=BAD_REQUEST,
+        )
+
+    _stop_server(server_pid)
+    _delete_server_state(server_state_file)
+
+    time.sleep(2)
```

Comment:
Is this necessary?

Reply:
not anymore! with the old way of using sigkill it was since there would be hanging processes. Now we're handling the termination a bit better it's completely unneeded. Thanks for the catch!